pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,584 @@
+			pubsub:   false,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Unsubscribe from a non-existent topic with an ID",
+			topic:    "h",
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Unsubscribe from an already unsubscribed topic with an ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   true,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Unsubscribe from a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientidd4",
+			err:      nil,
+			pubsub:   false,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Unsubscribe from an already unsubscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s.%s", chansPrefix, topic, subtopic),
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrNotSubscribed,
+			pubsub:   false,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			pubsub:   true,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Unsubscribe from an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid4",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			pubsub:   false,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to a topic with empty ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			pubsub:   true,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Unsubscribe from a topic with empty ID",
+			topic:    fmt.Sprintf("%s.%s", chansPrefix, topic),
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			pubsub:   false,
+			handler:  handler{false},
+		},
+	}
+	for _, tc := range cases {
+		switch tc.pubsub {
+		case true:
+			err := pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+			assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, tc.err, err))
+		default:
+			err := pubsub.Unsubscribe(tc.clientID, tc.topic)
+			assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, tc.err, err))
+		}
+	}
+}
+
+// Test Subscribe, Unsubscribe and Publish functions together (Test Entire PubSub interface)
+func TestAll(t *testing.T) {
```

Comment:
Use this test to test only PubSub - Publisher and Subscriber compatibility. No need for Unsubscribe - it's covered in the previous test.